controllers/storageclassclaim/storageclassclaim_controller.go

```diff
@@ -92,9 +109,153 @@ func (r *StorageClassClaimReconciler) Reconcile(ctx context.Context, request rec
+			Status:  corev1.ConditionTrue,
+			Reason:  "StorageClassClaim fulfilled and acknowledged.",
+			Message: "StorageClassClaim fulfilled and acknowledgement received from provider server.",
+		})
+		statusErr := r.Client.Status().Update(ctx, storageClassClaim)
+		if statusErr != nil {
+			log.Error(statusErr, "Failed to update StorageClassClaim status.")
+		}
+
+		/*	TODO
+			Periodically poll the provider server for resources provisioned for that claim.
+				Synchronous RPC since we need to wait for ACK to continue further
+				rpc PollProviderServer (PollProviderRequest) returns (PollProviderServerResponse) {}
+			Create (or do nothing if already created) the resources received from the provider server on the consumer cluster.
+		*/
+
+		// Create the StorageClass (StorageClass.Name == StorageClassClaim.Name) with StorageClassClaim as its owner.
+		_, err = controllerutil.CreateOrUpdate(ctx, r.Client, storageClass, func() error {
+			// storageClass with the same name as storageClassClaim is controlled by the storageClassClaim.
+			// storageClass must be garbage collected when the respective storageClassClaim is deleted.
+			return controllerutil.SetControllerReference(storageClassClaim, storageClass, r.Scheme)
+		})
+
+		// Update the StorageClassClaim status.
+		storageClassClaim.Status.Conditions = append(storageClassClaim.Status.Conditions, conditionsv1.Condition{
+			Type:    conditionsv1.ConditionAvailable,
+			Status:  corev1.ConditionTrue,
+			Reason:  "Resource creation successful.",
+			Message: "Created provider server resources and StorageClass on consumer cluster.",
+		})
+		statusErr = r.Client.Status().Update(ctx, storageClassClaim)
+		if statusErr != nil {
+			log.Error(statusErr, "Failed to update StorageClassClaim status.")
+		}
+
+	} else {
+
+		// Update the StorageClassClaim status.
+		storageClassClaim.Status.Conditions = append(storageClassClaim.Status.Conditions, conditionsv1.Condition{
+			Type:    conditionsv1.ConditionProgressing,
+			Status:  corev1.ConditionTrue,
+			Reason:  "Resource deletion initiated.",
+			Message: "Deleting managed resources.",
+		})
+		statusErr := r.Client.Status().Update(ctx, storageClassClaim)
+		if statusErr != nil {
+			log.Error(statusErr, "Failed to update StorageClassClaim status.")
+		}
+
+		// Wait until all PVs using the StorageClass under deletion are removed.
+		// Check for any PVs using the StorageClass.
+		pvList := &corev1.PersistentVolumeList{}
+		err := r.Client.List(ctx, pvList)
+		if err != nil {
+			return reconcile.Result{}, err
+		}
+		for _, pv := range pvList.Items {
+			if pv.Spec.StorageClassName == storageClassClaim.Name {
+				return reconcile.Result{}, fmt.Errorf("StorageClass %s is still in use by PV %s", storageClassClaim.Name, pv.Name)
+			}
+		}
+
+		// Delete StorageClass.
+		if err = r.get(ctx, storageClass); err == nil {
+			if err = r.delete(ctx, storageClass); err != nil {
+				return reconcile.Result{}, err
```

Comment:
Can we please add some more context to the error by wrapping it with our own error?